src/lib/data/history.ts

```diff
@@ -0,0 +1,87 @@
+import {db, Q, HistoryModel, RideModel} from '../../database';
+import {
+  TrainingCenterDatabase,
+  Activity,
+  ActivityLap,
+  Track,
+  TrackPoint,
+  Position,
+  HeartRateBpm,
+  TrackPointExtensions,
+  ActivityList,
+  HeartRateInBeatsPerMinute,
+} from 'tcx-builder';
+import {getRideAggregates} from './aggregates';
+
+export async function getRideHistory(ride: RideModel) {
+  return db.get<HistoryModel>('history').query(Q.where('ride_id', ride.id));
+}
+
+export async function generateTCX(ride: RideModel) {
+  if (!ride.endedAt) {
+    throw new Error(
+      `Ride ${ride.id} is not complete - can't generate tcx file.`,
+    );
+  }
+
+  const records = await getRideHistory(ride);
+  const aggregates = await getRideAggregates(ride);
+
+  const trackPoints: TrackPoint[] = [];
+  for (const record of records) {
+    trackPoints.push(
+      new TrackPoint({
+        time: new Date(record.createdAt),
+        ...(record.latitude &&
+          record.longitude && {
+            position: new Position(record.latitude, record.longitude),
+          }),
+        ...(record.altitude && {altitudeMeters: record.altitude}),
+        // not sure if they want distance between points or total
```

Comment:
I assumed it was total distance, but i could see how between points would be useful. I tried looking in the spec for some details but it doesn't really have any.

This is what GPT-4 says:

```
In the TCX (Training Center XML) file format, a TrackPoint object represents a single point along a GPS track with associated data. The distanceMeters field within a TrackPoint object represents the cumulative distance in meters from the starting point of the track to the current TrackPoint.
```

so it seems its the full accumulated distance

Reply:
Oh great thanks for checking :+1: